In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [54]:
a = pd.read_excel('recommender_data.xlsx')
a.head()

KeyboardInterrupt: 

In [3]:
a['emotion max'].value_counts()

['행복한']                                159669
['잔잔한']                                 79734
['경쾌한']                                 35646
['우울한']                                 25704
['행복한', '잔잔한']                          15265
                                        ...  
['우울한', '울고싶은', '행복한', '잔잔한']               1
['행복한', '잔잔한', '울고싶은', '우울한']               1
['잔잔한', '행복한', '무서운', '우울한', '경쾌한']         1
['경쾌한', '행복한', '긴장되는', '편안한']               1
['긴장되는', '무서운', '잔잔한', '경쾌한']               1
Name: emotion max, Length: 758, dtype: int64

In [4]:
sad = []
happy = []
peace = []
light = []
scary = []
comfort = []
cry = []
nervous = []

sad_to_csv = []
happy_to_csv = []
peace_to_csv = []
light_to_csv = []
scary_to_csv = []
comfort_to_csv = []
cry_to_csv = []
nervous_to_csv = []

for i in range(len(a)):
    if '우울한' in a['emotion max'][i]:
        sad.append(a['song id'][i])
        sad_to_csv.append(i)
    if '행복한' in a['emotion max'][i]:
        happy.append(a['song id'][i])
        happy_to_csv.append(i)
    if '잔잔한' in a['emotion max'][i]:
        peace.append(a['song id'][i])
        peace_to_csv.append(i)
    if '경쾌한' in a['emotion max'][i]:
        light.append(a['song id'][i])
        light_to_csv.append(i)
    if '무서운' in a['emotion max'][i]:
        scary.append(a['song id'][i])
        scary_to_csv.append(i)
    if '편안한' in a['emotion max'][i]:
        comfort.append(a['song id'][i])
        comfort_to_csv.append(i)
    if '울고싶은' in a['emotion max'][i]:
        cry.append(a['song id'][i])
        cry_to_csv.append(i)
    if '긴장되는' in a['emotion max'][i]:
        nervous.append(a['song id'][i])
        nervous_to_csv.append(i)

In [312]:
a.iloc[sad_to_csv].to_csv('sad_songs.csv', encoding='utf-8-sig')
a.iloc[happy_to_csv].to_csv('happy_songs.csv', encoding='utf-8-sig')
a.iloc[peace_to_csv].to_csv('peace_songs.csv', encoding='utf-8-sig')
a.iloc[light_to_csv].to_csv('light_songs.csv', encoding='utf-8-sig')
a.iloc[scary_to_csv].to_csv('scary_songs.csv', encoding='utf-8-sig')
a.iloc[comfort_to_csv].to_csv('comfort_songs.csv', encoding='utf-8-sig')
a.iloc[cry_to_csv].to_csv('cry_songs.csv', encoding='utf-8-sig')
a.iloc[nervous_to_csv].to_csv('nervous_songs.csv', encoding='utf-8-sig')

In [314]:
a.iloc[nervous_to_csv]

song id                                          tag count  \
11           18                                        {'스트레스': 1}   
18           34                                        {'스트레스': 1}   
22           40                     {'기분전환': 1, '분위기': 1, '섹시': 1}   
42           81                                        {'스트레스': 1}   
67          127  {'겨울': 1, '스트레스': 1, '출근': 1, '드라이브': 1, '퇴근':...   
...         ...                                                ...   
395226   707729                               {'스트레스': 1, '밴드': 1}   
395250   707774                             {'스트레스': 1, '드라이브': 1}   
395272   707809                      {'추억': 1, '스트레스': 1, '회상': 1}   
395331   707926  {'까페': 1, '분위기': 1, '달콤': 1, '기분전환': 2, '스트레스'...   
395337   707938                                        {'스트레스': 1}   

                                        tag max  \
11                                     ['스트레스']   
18                                     ['스트레스']   
22                        ['기분전환', '분위기', '섹시']   
42                                     ['스트레스']   
67      ['겨울', '스트레스', '출근', '드라이브', '퇴근', '집']   
...                                         ...   
395226                           ['스트레스', '밴드']   
395250                         ['스트레스', '드라이브']   
395272                     ['추억', '스트레스', '회상']   
395331                         ['기분전환', '스트레스']   
395337                                 ['스트레스']   

                                    emotion count             emotion max  
11                                    {'긴장되는': 1}                ['긴장되는']  
18                                    {'긴장되는': 1}                ['긴장되는']  
22                {'행복한': 1, '잔잔한': 1, '긴장되는': 1}  ['행복한', '잔잔한', '긴장되는']  
42                                    {'긴장되는': 1}                ['긴장되는']  
67      {'잔잔한': 1, '긴장되는': 2, '행복한': 1, '편안한': 2}         ['긴장되는', '편안한']  
...                                           ...                     ...  
395226                      {'긴장되는': 1, '경쾌한': 1}         ['긴장되는', '경쾌한']  
395250                      {'긴장되는': 1, '행복한': 1}         ['긴장되는', '행복한']  
395272            {'행복한': 1, '긴장되는': 1, '잔잔한': 1}  ['행복한', '긴장되는', '잔잔한']  
395331                      {'행복한': 1, '긴장되는': 1}         ['행복한', '긴장되는']  
395337                                {'긴장되는': 1}                ['긴장되는']  

[17299 rows x 5 columns]

In [43]:
a['song id'].values

array([     0,      3,      5, ..., 707982, 707986, 707987])

In [ ]:
train = pd.read_json('melon_music/train.json', typ = 'frame')
train.rename(columns={'tags' : '태그 리스트','id' : '플레이리스트 ID', 'plylst_title' : '플레이리스트 제목', 'songs' : '곡 리스트', 'like_cnt' : '좋아요 개수', 'updt_date' : '수정 날' }, inplace=True)
train.head()

In [ ]:
def filter_songs(songs):
    filtered_songs = []
    for i in range(len(songs)):
        if songs[i] in a['song id'].values:
            filtered_songs.append(songs[i])
    return filtered_songs

train['filter_songs'] = train['곡 리스트'].apply(lambda x: filter_songs(x))

In [ ]:
train.head()

In [ ]:
train.shape

In [ ]:
idx_list = []

for i in range(len(train)):
    if len(train['filter_songs'][i]) >= 5:
        idx_list.append(i)

In [ ]:
df = train.iloc[idx_list]
#df.to_csv('genre_filtered_songs.csv', encoding='utf-8-sig')

In [55]:
df = pd.read_csv('genre_filtered_songs.csv')
df.drop('Unnamed: 0', axis=1, inplace=True)
print(df.shape)

(112183, 7)


In [56]:
from sklearn.model_selection import train_test_split

train_df, test_df = train_test_split(df, test_size=0.2, random_state=42)

In [57]:
print(train_df.shape)
print(test_df.shape)

(89746, 7)
(22437, 7)


In [58]:
import fasttext
import fasttext.util
#fasttext.util.download_model('ko', if_exists='ignore')
ft = fasttext.load_model('../cc.ko.300.bin')

In [59]:
train_df.reset_index(inplace=True, drop=True)
test_df.reset_index(inplace=True, drop=True)

In [60]:
from tqdm import tqdm

# title
train_df['tag_vec_ft'] = '0'

vec = []
for i in tqdm(range(len(train_df))):
    train_df['태그 리스트'][i] = str(train_df['태그 리스트'][i])
    train_df['태그 리스트'][i] = train_df['태그 리스트'][i].replace('[', '').replace(']', '').replace("'", '')
    train_df['tag_vec_ft'][i] = ft.get_sentence_vector(train_df['태그 리스트'][i]).reshape((1,-1))

100%|█████████████████████████████████████| 89746/89746 [19:34<00:00, 76.38it/s]


In [62]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89746 entries, 0 to 89745
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   태그 리스트        89746 non-null  object
 1   플레이리스트 ID     89746 non-null  int64 
 2   플레이리스트 제목     89746 non-null  object
 3   곡 리스트         89746 non-null  object
 4   좋아요 개수        89746 non-null  int64 
 5   수정 날          89746 non-null  object
 6   filter_songs  89746 non-null  object
 7   tag_vec_ft    89746 non-null  object
dtypes: int64(2), object(6)
memory usage: 5.5+ MB


In [70]:
from tqdm import tqdm

# title
test_df['tag_vec_ft'] = '0'

vec = []
for i in tqdm(range(len(test_df))):
    test_df['태그 리스트'][i] = str(test_df['태그 리스트'][i])
    test_df['태그 리스트'][i] = test_df['태그 리스트'][i].replace('[', '').replace(']', '').replace("'", '')
    test_df['tag_vec_ft'][i] = ft.get_sentence_vector(test_df['태그 리스트'][i]).reshape((1,-1))

/var/folders/2d/k1h9bjt9385d884qrqvzk1880000gn/T/ipykernel_29919/317144339.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_df['tag_vec_ft'] = '0'
  0%|                                                 | 0/22437 [00:00<?, ?it/s]/var/folders/2d/k1h9bjt9385d884qrqvzk1880000gn/T/ipykernel_29919/317144339.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_df['태그 리스트'][i] = str(test_df['태그 리스트'][i])
/var/folders/2d/k1h9bjt9385d884qrqvzk1880000gn/T/ipykernel_29919/317144339.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice fro

In [71]:
#train_df.to_csv('processed_fast_train.csv', encoding='utf-8-sig')
#test_df.to_csv('processed_fast_test.csv', encoding='utf-8-sig')

In [76]:
print(len(test_df))
new_test_df = test_df.head(100)
print(len(new_test_df))

22437
100


In [77]:
## 추천 구하기
import warnings
warnings.filterwarnings('ignore')
from sklearn.metrics.pairwise import cosine_similarity

rec_result = []

for j in tqdm(range(len(new_test_df))):
    target = new_test_df['tag_vec_ft'].values[j][0].reshape(1, -1)
    
    train_df['cos_sim'] = train_df['tag_vec_ft'].apply(lambda _df: cosine_similarity(_df, target))

    rec_result.append((train_df.iloc[np.argmax(train_df['cos_sim'])]['곡 리스트'],
                      np.argmax(train_df['cos_sim']), max(train_df['cos_sim'])))
    

100%|█████████████████████████████████████████| 100/100 [46:06<00:00, 27.67s/it]


In [78]:
# 평가하기
class Evaluator:
    def _idcg(self, l):
        return sum((1.0 / np.log(i + 2) for i in range(l)))

    def __init__(self):
        self._idcgs = [self._idcg(i) for i in range(101)]

    def _ndcg(self, gt, rec):
        dcg = 0.0
        for i, r in enumerate(rec):
            if r in gt:
                dcg += 1.0 / np.log(i + 2)
        #print(len(gt))
        #print(len(rec))
        #print()
        return dcg / self._idcgs[len(gt)]
    
evaluator = Evaluator()

In [95]:
ndcg_total_score = []

for i in tqdm(range(len(new_test_df))):
    num = i

    true_relevance = new_test_df['곡 리스트'][num][:100]
    scores = rec_result[num][0][:100]
            
    ndcg_total_score.append(evaluator._ndcg(true_relevance, scores))    
print(np.mean(ndcg_total_score))

100%|███████████████████████████████████████| 100/100 [00:00<00:00, 3176.32it/s]

1.003245176529229


In [148]:
#### 무시 ~~~~~
ndcg_total_score = []

for i in tqdm(range(len(new_test_df))):
    num = i

    true_relevance = new_test_df['곡 리스트'][num].replace('[', '').replace(']', '').replace("'", '').replace(" ", '').split(',')[:100]
    scores = rec_result[num][0].replace('[', '').replace(']', '').replace("'", '').replace(" ", '').split(',')[:100]
            
    ndcg_total_score.append(evaluator._ndcg(true_relevance, scores))    
print(np.mean(ndcg_total_score))

  1%|▍                                         | 1/100 [00:00<00:00, 941.69it/s]


IndexError: list index out of range

#### 결론: # pre-trained fasttext를 이용한 임베딩시, test set 100개의 ndcg 평균: 1.003

## 유저 관점 테스트

In [3]:
song = pd.read_json('melon_music/song_meta.json',typ = 'frame')
song.rename(columns={"song_gn_dtl_gnr_basket":"곡 세부 장르 리스트", "issue_date":"발매일", "album_name": "앨범 명", "album_id":"앨범 ID", "artist_id_basket":"아티스트 ID 리스트", "song_name":"곡 제목", "song_gn_gnr_basket":"곡 장르 리스트", "artist_name_basket":"아티스트 리스트","id":"곡 ID"},inplace= True )
song.head()

곡 세부 장르 리스트       발매일  \
0          [GN0901]  20140512   
1  [GN1601, GN1606]  20080421   
2          [GN0901]  20180518   
3  [GN1102, GN1101]  20151016   
4  [GN1802, GN1801]  20110824   

                                                앨범 명    앨범 ID 아티스트 ID 리스트  \
0                       불후의 명곡 - 7080 추억의 얄개시대 팝송베스트  2255639      [2727]   
1                      Bach : Partitas Nos. 2, 3 & 4   376431     [29966]   
2                                                Hit  4698747      [3361]   
3  Feeling Right (Everything Is Nice) (Feat. Popc...  2644882    [838543]   
4                                            그남자 그여자  2008470    [560160]   

                                                곡 제목  곡 장르 리스트  \
0                                           Feelings  [GN0900]   
1  Bach : Partita No. 4 In D Major, BWV 828 - II....  [GN1600]   
2                    Solsbury Hill (Remastered 2002)  [GN0900]   
3  Feeling Right (Everything Is Nice) (Feat. Popc...  [GN1100]   
4                                            그남자 그여자  [GN1800]   

            아티스트 리스트  곡 ID  
0  [Various Artists]     0  
1   [Murray Perahia]     1  
2    [Peter Gabriel]     2  
3           [Matoma]     3  
4         [Jude Law]     4

In [4]:
songs_with_tags = pd.merge(song, a, left_on='곡 ID', right_on = 'song id', how='outer')#['song id']
songs_with_tags = songs_with_tags.set_index('곡 ID', drop=True)

In [ ]:
train = pd.read_json('melon_music/train.json', typ = 'frame')
train.rename(columns={'tags' : '태그 리스트','id' : '플레이리스트 ID', 'plylst_title' : '플레이리스트 제목', 'songs' : '곡 리스트', 'like_cnt' : '좋아요 개수', 'updt_date' : '수정 날' }, inplace=True)
train.head()

In [235]:
# 유저 관점:

user_tag = '힐링, 쉼, 휴식, 고요'
emotion_result = ('잔잔한', 0.7), ('편안한', 0.3)
top_n = 5
max_song_list = 100
################

emotion_n = len(emotion_result)

sa_active = 0
h_active = 0
p_active = 0
l_active = 0
sc_active = 0
co_active = 0
cr_active = 0
n_active = 0

for emo_idx in range(len(emotion_result)):
    emo = emotion_result[emo_idx][0]
    if emo == '잔잔한':
        p_active = emotion_result[emo_idx][1]
    if emo == '편안한':
        co_active = emotion_result[emo_idx][1]
    if emo == '우울한':
        sa_active = emotion_result[emo_idx][1]
    if emo == '행복한':
        h_active = emotion_result[emo_idx][1]
    if emo == '경쾌한':
        l_active = emotion_result[emo_idx][1]
    if emo == '무서운':
        sc_active = emotion_result[emo_idx][1]
    if emo == '울고싶은':
        cr_active = emotion_result[emo_idx][1]
    if emo == '긴장되는':
        n_active = emotion_result[emo_idx][1]

In [236]:
user_tag_vectors = ft.get_sentence_vector(user_tag).reshape((1,-1))

## 추천 구하기
import warnings
warnings.filterwarnings('ignore')
from sklearn.metrics.pairwise import cosine_similarity

rec_result = []
    
train_df['cos_sim'] = train_df['tag_vec_ft'].apply(lambda _df: cosine_similarity(_df, user_tag_vectors))
## 시간 오래걸리는데 train 사이즈를 줄일까?

In [238]:
a[a['song id'] == 96310]


song id                                          tag count tag max  \
53528    96310  {'햇살': 4, '봄': 49, '카페': 80, '봄바람': 6, '주말': 2...  ['사랑']   

      emotion count emotion max  
53528    {'행복한': 1}     ['행복한']

In [316]:
import heapq
fr = train_df['cos_sim']
n_result = heapq.nlargest(max_song_list, range(len(fr)), key=fr.__getitem__)

print('active 값:', sa_active, h_active, p_active, l_active, sc_active, co_active, cr_active, n_active)

found_list =[]
found = 0
flag = 1

for i in range(max_song_list):
    sc_list = []
    rec_result = []
    rec_result.append((train_df.iloc[n_result[i]]['곡 리스트'],
                   train_df.iloc[n_result[i]]['cos_sim']))

    user_rec_songs = rec_result[0][0].replace('[', '').replace(']', '').replace("'", '').replace(" ", '').split(',')#[:100]
    user_rec_songs = list(map(int, user_rec_songs))
    print('similarity', rec_result[0][1])

    for song in user_rec_songs:
        sc = 0
        if sa_active:
            if song in sad:
                sc += sa_active
        if h_active:
            if song in happy:
                sc += h_active
        if p_active:
            if song in peace:
                sc += p_active
        if l_active:
            if song in light:
                sc += l_active
        if sc_active:
            if song in scary:
                sc += sc_active
        if co_active:
            if song in comfort:
                sc += co_active
        if cr_active:
            if song in cry:
                sc += cr_active
        if n_active:
            if song in nerbous:
                sc += n_active
        
        if sc>0:
            print('song:', song, 'sc:', sc)
            sc_list.append(sc)

            found += 1
            found_list.append(song)
        
    if found >= top_n:
        break


active 값: 0 0 0.7 0 0 0.3 0 0
similarity [[0.83917713]]
song: 604870 sc: 0.7
song: 480675 sc: 0.7
song: 698648 sc: 0.7
song: 125822 sc: 0.7
song: 431225 sc: 0.7


In [317]:
# 후보
found_list

[604870, 480675, 698648, 125822, 431225]

In [318]:
# sc 높은 순으로 추천
song_result = heapq.nlargest(top_n, range(len(sc_list)), key=sc_list.__getitem__)
fin_song_result = [found_list[song_result[i]] for i in range(top_n)]
fin_song_result

[604870, 480675, 698648, 125822, 431225]

In [324]:
song.iloc[fin_song_result]

곡 세부 장르 리스트       발매일  \
604870           [GN2502, GN0105, GN2501, GN0101, GN2505]  20151012   
480675                   [GN2502, GN0601, GN2501, GN0606]  20120710   
698648  [GN0105, GN0101, GN2505, GN2502, GN1501, GN250...  20161115   
125822           [GN0105, GN2505, GN2501, GN0101, GN2503]  20161123   
431225                                   [GN0105, GN0101]  20160911   

                                       앨범 명     앨범 ID  \
604870                               I Mean   2644202   
480675                              Goodbye   2136832   
698648                캐리어를 끄는 여자 OST Part.6  10015096   
125822  Jelly box 꽃길 (Prod. By 지코(ZICO)) 세정  10016726   
431225                             복면가왕 76회   2710626   

                             아티스트 ID 리스트                         곡 제목  \
604870  [672287, 672288, 672289, 672296]                       여기 있을게   
480675                          [623451]        Goodbye (With Matoki)   
698648                          [670611]                        인형의 꿈   
125822                          [968261]           꽃길 (Prod. By ZICO)   
431225                          [672287]  바람이 분다 (오늘 밤엔 어둠이 무서워요 석봉이)   

                        곡 장르 리스트                          아티스트 리스트    곡 ID  
604870          [GN2500, GN0100]  [서은광, 이창섭, 임현식 (비투비), 육성재 (비투비)]  604870  
480675          [GN2500, GN0600]                           [B.A.P]  480675  
698648  [GN2500, GN1500, GN0100]                        [태일 (블락비)]  698648  
125822          [GN2500, GN0100]                        [세정 (구구단)]  125822  
431225                  [GN0100]                             [서은광]  431225

In [325]:
songs_with_tags.iloc[fin_song_result]

곡 세부 장르 리스트       발매일  \
곡 ID                                                                  
604870           [GN2502, GN0105, GN2501, GN0101, GN2505]  20151012   
480675                   [GN2502, GN0601, GN2501, GN0606]  20120710   
698648  [GN0105, GN0101, GN2505, GN2502, GN1501, GN250...  20161115   
125822           [GN0105, GN2505, GN2501, GN0101, GN2503]  20161123   
431225                                   [GN0105, GN0101]  20160911   

                                       앨범 명     앨범 ID  \
곡 ID                                                    
604870                               I Mean   2644202   
480675                              Goodbye   2136832   
698648                캐리어를 끄는 여자 OST Part.6  10015096   
125822  Jelly box 꽃길 (Prod. By 지코(ZICO)) 세정  10016726   
431225                             복면가왕 76회   2710626   

                             아티스트 ID 리스트                         곡 제목  \
곡 ID                                                                    
604870  [672287, 672288, 672289, 672296]                       여기 있을게   
480675                          [623451]        Goodbye (With Matoki)   
698648                          [670611]                        인형의 꿈   
125822                          [968261]           꽃길 (Prod. By ZICO)   
431225                          [672287]  바람이 분다 (오늘 밤엔 어둠이 무서워요 석봉이)   

                        곡 장르 리스트                          아티스트 리스트   song id  \
곡 ID                                                                           
604870          [GN2500, GN0100]  [서은광, 이창섭, 임현식 (비투비), 육성재 (비투비)]  604870.0   
480675          [GN2500, GN0600]                           [B.A.P]  480675.0   
698648  [GN2500, GN1500, GN0100]                        [태일 (블락비)]  698648.0   
125822          [GN2500, GN0100]                        [세정 (구구단)]  125822.0   
431225                  [GN0100]                             [서은광]  431225.0   

                                                tag count        tag max  \
곡 ID                                                                       
604870  {'따뜻한': 1, '사랑': 11, '이별': 6, '까페': 1, '힐링': 1...         ['휴식']   
480675  {'지친하루': 1, '위로': 5, '휴식': 12, '힐링': 11, '신나는'...         ['휴식']   
698648  {'힐링': 2, '휴식': 4, '봄': 2, '달달': 2, '커플': 1, '...         ['휴식']   
125822  {'후회': 4, '눈물': 19, '사랑': 75, '새벽': 60, '우울': ...  ['힐링', '잔잔한']   
431225  {'사랑': 2, '이별': 2, '댄스': 3, '휴식': 3, '눈오는날': 1...   ['댄스', '휴식']   

               emotion count     emotion max  
곡 ID                                          
604870            {'잔잔한': 1}         ['잔잔한']  
480675            {'잔잔한': 1}         ['잔잔한']  
698648            {'잔잔한': 1}         ['잔잔한']  
125822  {'행복한': 1, '잔잔한': 1}  ['행복한', '잔잔한']  
431225  {'경쾌한': 1, '잔잔한': 1}  ['경쾌한', '잔잔한']

In [303]:
355771 in peace

True

In [5]:
songs_with_tags.to_json('songs_with_tags.json', orient='table')

In [6]:
new_data = pd.read_json('songs_with_tags.json', orient='table')
new_data

곡 세부 장르 리스트       발매일  \
곡 ID                                 
0               [GN0901]  20140512   
1       [GN1601, GN1606]  20080421   
2               [GN0901]  20180518   
3       [GN1102, GN1101]  20151016   
4       [GN1802, GN1801]  20110824   
...                  ...       ...   
707984          [GN2001]  19991219   
707985          [GN0901]  19860000   
707986  [GN0105, GN0101]  20160120   
707987  [GN1807, GN1801]  20131217   
707988  [GN0601, GN0604]  19980000   

                                                     앨범 명    앨범 ID  \
곡 ID                                                                 
0                            불후의 명곡 - 7080 추억의 얄개시대 팝송베스트  2255639   
1                           Bach : Partitas Nos. 2, 3 & 4   376431   
2                                                     Hit  4698747   
3       Feeling Right (Everything Is Nice) (Feat. Popc...  2644882   
4                                                 그남자 그여자  2008470   
...                                                   ...      ...   
707984               The Best Best Of The Black President    65254   
707985                                        True Colors    44141   
707986                  행보 2015 윤종신 / 작사가 윤종신 Live Part.1  2662866   
707987                                 명상의 시간을 위한 뉴에이지 음악  2221722   
707988                                           김경호 Live    34663   

       아티스트 ID 리스트                                               곡 제목  \
곡 ID                                                                    
0           [2727]                                           Feelings   
1          [29966]  Bach : Partita No. 4 In D Major, BWV 828 - II....   
2           [3361]                    Solsbury Hill (Remastered 2002)   
3         [838543]  Feeling Right (Everything Is Nice) (Feat. Popc...   
4         [560160]                                            그남자 그여자   
...            ...                                                ...   
707984    [166499]                           Coffin For Head Of State   
707985     [11837]                                    Change Of Heart   
707986       [437]                                             스치듯 안녕   
707987    [729868]                                               숲의 빛   
707988       [895]                                       Queen 명곡 멜로디   

        곡 장르 리스트           아티스트 리스트   song id  \
곡 ID                                            
0       [GN0900]  [Various Artists]       0.0   
1       [GN1600]   [Murray Perahia]       NaN   
2       [GN0900]    [Peter Gabriel]       NaN   
3       [GN1100]           [Matoma]       3.0   
4       [GN1800]         [Jude Law]       NaN   
...          ...                ...       ...   
707984  [GN2000]        [Fela Kuti]       NaN   
707985  [GN0900]     [Cyndi Lauper]       NaN   
707986  [GN0100]              [윤종신]  707986.0   
707987  [GN1800]     [Nature Piano]  707987.0   
707988  [GN0600]              [김경호]       NaN   

                                                tag count  \
곡 ID                                                        
0                {'비오는날': 2, '드라이브': 1, '추억': 1, '회상': 1}   
1                                                     NaN   
2                                                     NaN   
3       {'드라이브': 2, '여행': 1, '기분전환': 3, '신나는': 3, '클럽'...   
4                                                     NaN   
...                                                   ...   
707984                                                NaN   
707985                                                NaN   
707986  {'카페': 1, '잔잔한': 4, '추억': 3, '위로': 2, '힐링': 1,...   
707987  {'집중': 1, '노동요': 1, '새벽': 1, '오후': 1, '자장가': 1...   
707988                                                NaN   

                                       tag max  \
곡 ID                                             
0                                     ['비오는날']   
1                                          NaN   
2                        

In [7]:
fin_song_result = [604870, 480675, 698648, 125822, 431225]

In [10]:
songs_with_tags.loc[fin_song_result, ['곡 제목', '아티스트 리스트', 'tag max', 'emotion max']]

곡 제목                          아티스트 리스트  \
곡 ID                                                                    
604870                       여기 있을게  [서은광, 이창섭, 임현식 (비투비), 육성재 (비투비)]   
480675        Goodbye (With Matoki)                           [B.A.P]   
698648                        인형의 꿈                        [태일 (블락비)]   
125822           꽃길 (Prod. By ZICO)                        [세정 (구구단)]   
431225  바람이 분다 (오늘 밤엔 어둠이 무서워요 석봉이)                             [서은광]   

              tag max     emotion max  
곡 ID                                   
604870         ['휴식']         ['잔잔한']  
480675         ['휴식']         ['잔잔한']  
698648         ['휴식']         ['잔잔한']  
125822  ['힐링', '잔잔한']  ['행복한', '잔잔한']  
431225   ['댄스', '휴식']  ['경쾌한', '잔잔한']

In [12]:
train_df = pd.read_csv('processed_fast_train.csv')
train_df.drop('Unnamed: 0', axis=1, inplace=True)

In [13]:
train_df.to_json('processed_fast_train.json', orient='table')

In [14]:
new_data = pd.read_json('processed_fast_train.json', orient='table')
new_data

태그 리스트  플레이리스트 ID  \
0      비오는날, 매장, 어쿠스틱, 카페, 새벽, 피아노, 연주곡, 가을, 뉴에이지, 감미로운      93926   
1                     퇴근, 밤, 음색, 젊음, 위로, 힐링, 희망, 청춘, 지친     108218   
2                                                   스무살       9244   
3                                               게임, OST      99381   
4                                           여름, 더위, 신나는      90542   
...                                                 ...        ...   
89741                                       힐링, 휴식, 잔잔한      70958   
89742                                                가을      41383   
89743                                            추억, 회상      70006   
89744                   매장음악, 운동, 신나는음악, 드라이브, 기분전환, 비트     114568   
89745                                             밴드, 락      53922   

                     플레이리스트 제목  \
0      새벽감성 센치한 느낌이 너무 좋은 연주곡들   
1        하루 끝, 지친 날 위로해준 힐링 음색   
2              오늘의 추천 가수 - 스무살   
3                   Game OST 2   
4                     여름아 부탁해!   
...                        ...   
89741                   The xx   
89742                 ㅇㅇㅇㅇㅇㅇㅇㅇ   
89743           그리움을 품은 쓸쓸한 음악   
89744           매장에 쓸수있는 신나는음악   
89745                   반오십 밴드   

                                                   곡 리스트  좋아요 개수  \
0      [293751, 574777, 451783, 446059, 335112, 30995...      22   
1      [672550, 132808, 415394, 368429, 96310, 630395...    2718   
2      [181217, 406731, 591700, 30599, 227036, 260822...      11   
3      [492350, 171668, 640828, 526752, 519752, 20703...       5   
4      [95059, 277034, 114387, 200687, 628491, 343974...       8   
...                                                  ...     ...   
89741  [514089, 433441, 98271, 190868, 326424, 557326...      18   
89742  [502709, 150349, 36918, 174849, 397237, 7749, ...      19   
89743  [670853, 92973, 595181, 664155, 453156, 140867...       0   
89744  [581086, 305728, 529877, 537733, 67874, 187526...      27   
89745  [424318, 439572, 566926, 581245, 28832, 338033...       1   

                          수정 날  \
0      2017-09-22 17:21:40.000   
1      2019-04-29 06:01:40.000   
2      2019-04-17 18:35:30.000   
3      2019-11-30 16:42:57.000   
4      2019-06-22 17:38:09.000   
...                        ...   
89741  2019-02-25 13:18:25.000   
89742  2016-09-13 22:32:34.000   
89743  2015-10-02 15:13:23.000   
89744  2017-12-18 01:54:17.000   
89745  2020-02-01 08:28:40.000   

                                            filter_songs  \
0      [293751, 574777, 451783, 446059, 335112, 30995...   
1      [672550, 132808, 415394, 368429, 96310, 630395...   
2      [181217, 406731, 591700, 30599, 227036, 260822...   
3      [492350, 171668, 640828, 526752, 519752, 20703...   
4      [95059, 277034, 114387, 200687, 628491, 343974...   
...                                                  ...   
89741  [514089, 433441, 98271, 190868, 326424, 557326...   
89742  [502709, 150349, 36918, 174849, 397237, 7749, ...   
89743  [670853, 92973, 595181, 664155, 453156, 140867...   
89744  [581086, 305728, 529877, 537733, 67874, 187526...   
89745  [424318, 439572, 566926, 581245, 28832, 338033...   

                                              tag_vec_ft  
0      [[-2.17173584e-02 -1.42692477e-02 -3.08524948e...  
1      [[-8.69084615e-03  4.53952421e-03 -1.50658088e...  
2      [[-1.79576762e-02 -2.31178794e-02 -2.11853441e...  
3      [[-5.15502617e-02 -4.54587042e-02 -1.54404528e...  
4      [[ 1.29450653e-02 -7.03520840e-03 -4.68925983e...  
...                                                  ...  
89741  [[-5.89748137e-02 -1.02304236e-03  8.22014641e...  
89742  [[-3.44714150e-02 -2.83767283e-02 -4.27337699e...  
89743  [[ 2.89065279e-02 -4.24530171e-02  7.45522603e...  
89744  [[-1.37153035e-02  3.44576947e-02  5.62024340e...  
89745  [[-2.78151762e-02 -5.90510257e-02 -4.62382473e...  

[89746 rows x 8 columns]

In [ ]:
import fasttext
import fasttext.util
#fasttext.util.download_model('ko', if_exists='ignore')
ft = fasttext.load_model('../cc.ko.300.bin')

In [16]:
train_df = pd.read_csv('processed_fast_train.csv')
train_df.drop('Unnamed: 0', axis=1, inplace=True)

In [63]:
from sklearn.metrics.pairwise import cosine_similarity

user_tag = '힐링, 쉼, 휴식, 고요'
user_tag_vectors = ft.get_sentence_vector(user_tag).reshape((1,-1))


In [25]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89746 entries, 0 to 89745
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   태그 리스트        89746 non-null  object
 1   플레이리스트 ID     89746 non-null  int64 
 2   플레이리스트 제목     89746 non-null  object
 3   곡 리스트         89746 non-null  object
 4   좋아요 개수        89746 non-null  int64 
 5   수정 날          89746 non-null  object
 6   filter_songs  89746 non-null  object
 7   tag_vec_ft    89746 non-null  object
dtypes: int64(2), object(6)
memory usage: 5.5+ MB


In [66]:
train_df.to_csv('pp_train.csv', encoding='utf-8-sig')

In [67]:
train_df.to_json('pp_train.json', orient='table')

In [68]:
a = pd.read_csv('pp_train.csv')
a.drop('Unnamed: 0', axis=1)


태그 리스트  플레이리스트 ID  \
0      비오는날, 매장, 어쿠스틱, 카페, 새벽, 피아노, 연주곡, 가을, 뉴에이지, 감미로운      93926   
1                     퇴근, 밤, 음색, 젊음, 위로, 힐링, 희망, 청춘, 지친     108218   
2                                                   스무살       9244   
3                                               게임, OST      99381   
4                                           여름, 더위, 신나는      90542   
...                                                 ...        ...   
89741                                       힐링, 휴식, 잔잔한      70958   
89742                                                가을      41383   
89743                                            추억, 회상      70006   
89744                   매장음악, 운동, 신나는음악, 드라이브, 기분전환, 비트     114568   
89745                                             밴드, 락      53922   

                     플레이리스트 제목  \
0      새벽감성 센치한 느낌이 너무 좋은 연주곡들   
1        하루 끝, 지친 날 위로해준 힐링 음색   
2              오늘의 추천 가수 - 스무살   
3                   Game OST 2   
4                     여름아 부탁해!   
...                        ...   
89741                   The xx   
89742                 ㅇㅇㅇㅇㅇㅇㅇㅇ   
89743           그리움을 품은 쓸쓸한 음악   
89744           매장에 쓸수있는 신나는음악   
89745                   반오십 밴드   

                                                   곡 리스트  좋아요 개수  \
0      [293751, 574777, 451783, 446059, 335112, 30995...      22   
1      [672550, 132808, 415394, 368429, 96310, 630395...    2718   
2      [181217, 406731, 591700, 30599, 227036, 260822...      11   
3      [492350, 171668, 640828, 526752, 519752, 20703...       5   
4      [95059, 277034, 114387, 200687, 628491, 343974...       8   
...                                                  ...     ...   
89741  [514089, 433441, 98271, 190868, 326424, 557326...      18   
89742  [502709, 150349, 36918, 174849, 397237, 7749, ...      19   
89743  [670853, 92973, 595181, 664155, 453156, 140867...       0   
89744  [581086, 305728, 529877, 537733, 67874, 187526...      27   
89745  [424318, 439572, 566926, 581245, 28832, 338033...       1   

                          수정 날  \
0      2017-09-22 17:21:40.000   
1      2019-04-29 06:01:40.000   
2      2019-04-17 18:35:30.000   
3      2019-11-30 16:42:57.000   
4      2019-06-22 17:38:09.000   
...                        ...   
89741  2019-02-25 13:18:25.000   
89742  2016-09-13 22:32:34.000   
89743  2015-10-02 15:13:23.000   
89744  2017-12-18 01:54:17.000   
89745  2020-02-01 08:28:40.000   

                                            filter_songs  \
0      [293751, 574777, 451783, 446059, 335112, 30995...   
1      [672550, 132808, 415394, 368429, 96310, 630395...   
2      [181217, 406731, 591700, 30599, 227036, 260822...   
3      [492350, 171668, 640828, 526752, 519752, 20703...   
4      [95059, 277034, 114387, 200687, 628491, 343974...   
...                                                  ...   
89741  [514089, 433441, 98271, 190868, 326424, 557326...   
89742  [502709, 150349, 36918, 174849, 397237, 7749, ...   
89743  [670853, 92973, 595181, 664155, 453156, 140867...   
89744  [581086, 305728, 529877, 537733, 67874, 187526...   
89745  [424318, 439572, 566926, 581245, 28832, 338033...   

                                              tag_vec_ft  
0      [[-2.17173584e-02 -1.42692477e-02 -3.08524948e...  
1      [[-8.69084615e-03  4.53952421e-03 -1.50658088e...  
2      [[-1.79576762e-02 -2.31178794e-02 -2.11853441e...  
3      [[-5.15502617e-02 -4.54587042e-02 -1.54404528e...  
4      [[ 1.29450653e-02 -7.03520840e-03 -4.68925983e...  
...                                                  ...  
89741  [[-5.89748137e-02 -1.02304236e-03  8.22014641e...  
89742  [[-3.44714150e-02 -2.83767283e-02 -4.27337699e...  
89743  [[ 2.89065279e-02 -4.24530171e-02  7.45522603e...  
89744  [[-1.37153035e-02  3.44576947e-02  5.62024340e...  
89745  [[-2.78151762e-02 -5.90510257e-02 -4.62382473e...  

[89746 rows x 8 columns]

In [83]:
#song = pd.read_json('pp_train.json')#,typ = 'frame')
import json
import pandas as pd
data = json.load(open('pp_train.json'))

df = pd.DataFrame(data["data"])

In [86]:
df["tag_vec_ft"]

0        [[-0.0217173584, -0.0142692477, -0.0308524948,...
1        [[-0.0086908462, 0.0045395242, -0.0150658088, ...
2        [[-0.0179576762, -0.0231178794, -0.0211853441,...
3        [[-0.0515502617, -0.0454587042, -0.0015440453,...
4        [[0.0129450653, -0.0070352084, -0.0468925983, ...
                               ...                        
89741    [[-0.0589748137, -0.0010230424, 0.0082201464, ...
89742    [[-0.034471415, -0.0283767283, -0.0427337699, ...
89743    [[0.0289065279, -0.0424530171, 0.007455226, -0...
89744    [[-0.0137153035, 0.0344576947, 0.056202434, 0....
89745    [[-0.0278151762, -0.0590510257, -0.0046238247,...
Name: tag_vec_ft, Length: 89746, dtype: object

In [87]:
df['cos_sim'] = df['tag_vec_ft'].apply(lambda _df: cosine_similarity(_df, user_tag_vectors))



In [88]:
df

index                                            태그 리스트  플레이리스트 ID  \
0          0  비오는날, 매장, 어쿠스틱, 카페, 새벽, 피아노, 연주곡, 가을, 뉴에이지, 감미로운      93926   
1          1                 퇴근, 밤, 음색, 젊음, 위로, 힐링, 희망, 청춘, 지친     108218   
2          2                                               스무살       9244   
3          3                                           게임, OST      99381   
4          4                                       여름, 더위, 신나는      90542   
...      ...                                               ...        ...   
89741  89741                                       힐링, 휴식, 잔잔한      70958   
89742  89742                                                가을      41383   
89743  89743                                            추억, 회상      70006   
89744  89744                   매장음악, 운동, 신나는음악, 드라이브, 기분전환, 비트     114568   
89745  89745                                             밴드, 락      53922   

                     플레이리스트 제목  \
0      새벽감성 센치한 느낌이 너무 좋은 연주곡들   
1        하루 끝, 지친 날 위로해준 힐링 음색   
2              오늘의 추천 가수 - 스무살   
3                   Game OST 2   
4                     여름아 부탁해!   
...                        ...   
89741                   The xx   
89742                 ㅇㅇㅇㅇㅇㅇㅇㅇ   
89743           그리움을 품은 쓸쓸한 음악   
89744           매장에 쓸수있는 신나는음악   
89745                   반오십 밴드   

                                                   곡 리스트  좋아요 개수  \
0      [293751, 574777, 451783, 446059, 335112, 30995...      22   
1      [672550, 132808, 415394, 368429, 96310, 630395...    2718   
2      [181217, 406731, 591700, 30599, 227036, 260822...      11   
3      [492350, 171668, 640828, 526752, 519752, 20703...       5   
4      [95059, 277034, 114387, 200687, 628491, 343974...       8   
...                                                  ...     ...   
89741  [514089, 433441, 98271, 190868, 326424, 557326...      18   
89742  [502709, 150349, 36918, 174849, 397237, 7749, ...      19   
89743  [670853, 92973, 595181, 664155, 453156, 140867...       0   
89744  [581086, 305728, 529877, 537733, 67874, 187526...      27   
89745  [424318, 439572, 566926, 581245, 28832, 338033...       1   

                          수정 날  \
0      2017-09-22 17:21:40.000   
1      2019-04-29 06:01:40.000   
2      2019-04-17 18:35:30.000   
3      2019-11-30 16:42:57.000   
4      2019-06-22 17:38:09.000   
...                        ...   
89741  2019-02-25 13:18:25.000   
89742  2016-09-13 22:32:34.000   
89743  2015-10-02 15:13:23.000   
89744  2017-12-18 01:54:17.000   
89745  2020-02-01 08:28:40.000   

                                            filter_songs  \
0      [293751, 574777, 451783, 446059, 335112, 30995...   
1      [672550, 132808, 415394, 368429, 96310, 630395...   
2      [181217, 406731, 591700, 30599, 227036, 260822...   
3      [492350, 171668, 640828, 526752, 519752, 20703...   
4      [95059, 277034, 114387, 200687, 628491, 343974...   
...                                                  ...   
89741  [514089, 433441, 98271, 190868, 326424, 557326...   
89742  [502709, 150349, 36918, 174849, 397237, 7749, ...   
89743  [670853, 92973, 595181, 664155, 453156, 140867...   
89744  [581086, 305728, 529877, 537733, 67874, 187526...   
89745  [424318, 439572, 566926, 581245, 28832, 338033...   

                                              tag_vec_ft  \
0      [[-0.0217173584, -0.0142692477, -0.0308524948,...   
1      [[-0.0086908462, 0.0045395242, -0.0150658088, ...   
2      [[-0.0179576762, -0.0231178794, -0.0211853441,...   
3      [[-0.0515502617, -0.0454587042, -0.0015440453,...   
4      [[0.0129450653, -0.0070352084, -0.0468925983, ...   
...                                                  ...   
89741  [[-0.0589748137, -0.0010230424, 0.0082201464, ...   
89742  [[-0.034471415, -0.0283767283, -0.0427337699, ...   
89743  [[0.0289065279, -0.0424530171, 0.007455226, -0...   
89744  [[-0.0137153035, 0.0344576947, 0.056202434, 0....   
89745  [[-0.0278151762, -0.0590510257, -0.0046238247,...   

     

In [71]:
a['cos_sim'] = a['tag_vec_ft'].apply(lambda _df: cosine_similarity(_df, user_tag_vectors))


ValueError: could not convert string to float: '[[-2.17173584e-02 -1.42692477e-02 -3.08524948e-02 -9.46905639e-04\n   1.38190640e-02 -2.20028739e-02  2.43856218e-02 -1.20863346e-02\n  -8.69168900e-03 -3.30825709e-03 -2.94765886e-02  1.25598907e-02\n  -2.09537987e-02  1.71110854e-02  7.05836061e-03  8.95220228e-03\n  -2.05893232e-03  1.32553941e-02  2.85860095e-02 -1.77105889e-04\n  -1.13778384e-02 -6.55439857e-04 -9.87571385e-03  2.91636325e-02\n  -1.15482723e-02  1.23493545e-01 -1.03799021e-02 -2.22634673e-02\n  -1.06921531e-02  2.27064695e-02  5.05730230e-03  2.49180961e-02\n  -8.38364940e-03  9.06019565e-03  1.31057575e-02  4.46712561e-02\n  -3.16444151e-02  5.17214695e-03  4.21260251e-03 -5.84957015e-04\n   1.58183482e-02 -8.88033807e-02  1.64859127e-02  1.19223469e-03\n   4.39577661e-02  6.18726499e-02 -4.71269526e-03  9.97465570e-03\n   2.24705525e-02 -1.59197710e-02  2.07750443e-02  2.53249444e-02\n  -3.73518243e-02 -3.45849432e-02  2.11226800e-03 -6.76189503e-03\n   4.53965329e-02 -1.72852464e-02 -7.60814324e-02 -6.71991194e-03\n   1.71274003e-02 -7.67265866e-03  2.73759644e-02 -1.47069143e-02\n   2.14122981e-02  2.58114990e-02 -2.02316861e-03 -1.27713261e-02\n  -4.69594076e-02 -5.17785810e-02 -1.28978193e-02 -3.05174626e-02\n  -4.83104549e-02  8.22246261e-03 -8.30200326e-04 -6.29269052e-03\n   7.96356332e-03  2.96143070e-03 -4.91908267e-02  3.19042131e-02\n  -2.89269630e-02 -6.18031263e-05  1.45858256e-02  3.93935107e-02\n  -3.41555215e-02 -1.53222159e-02 -4.27433327e-02 -4.71075177e-02\n   1.44415963e-02 -2.44041835e-03  3.48705659e-03 -3.11970413e-02\n  -4.16710041e-02  1.36679098e-01 -4.77851572e-04  6.95204129e-03\n   2.36228210e-04  8.15412961e-03  2.44333269e-03  9.20857023e-03\n  -1.08063212e-02 -4.62675793e-03 -2.40413677e-02  1.59284156e-02\n  -1.02366293e-02 -4.24863473e-02  1.35527942e-02  2.43274868e-02\n  -1.22906184e-02  3.86634818e-03  1.45013807e-02  2.58904155e-02\n  -6.56927517e-03 -1.23844594e-02 -2.97040585e-02  3.86172011e-02\n  -8.90891720e-03  2.71167662e-02  3.03208772e-02  4.68451232e-02\n   6.16739579e-02 -5.83716063e-03 -5.34775816e-02  1.76265016e-02\n   1.30194426e-03 -2.25897282e-02 -3.24477479e-02 -1.96740590e-02\n   1.65073983e-02 -2.24523782e-03 -1.02402633e-02 -2.93175615e-02\n   8.20101574e-02  7.48990625e-02  7.78166205e-03 -3.98168191e-02\n  -1.89875495e-02  1.20302634e-02 -6.14041789e-03  3.64032807e-03\n   1.55361118e-02 -3.89172845e-02 -1.22744022e-02  1.58310570e-02\n  -1.89588964e-02 -3.57277580e-02 -2.76223253e-02 -2.79105660e-02\n   3.57055967e-03 -3.32635082e-02  2.64081750e-02 -5.35346568e-03\n  -8.32914375e-03 -1.62245054e-02  3.24964188e-02 -6.73158048e-03\n   1.64105895e-03 -8.99581909e-02 -2.66295504e-02  2.91870181e-02\n   8.35937168e-03 -1.82381608e-02 -1.51372096e-02 -1.79124065e-03\n  -9.99099482e-03 -9.07322019e-03  1.88689586e-02  7.76607683e-03\n   9.10605676e-03 -1.83197223e-02  1.52185336e-02 -1.62944403e-02\n  -3.16022262e-02  5.83689800e-03  1.40054394e-02  1.48241846e-02\n  -1.47276223e-02  3.88295390e-03 -3.01101264e-02 -5.51706878e-03\n  -2.00618207e-02  4.71663773e-02  2.63146106e-02  4.89660120e-03\n   2.22494286e-02 -2.97602993e-02 -1.32446913e-02 -2.78351996e-02\n   1.28366593e-02 -5.27260303e-02 -9.27329995e-04  6.71971664e-02\n   6.95215538e-03  9.10225231e-03 -4.51201951e-04  4.41013239e-02\n  -2.72892397e-02 -3.82483983e-03 -3.18574086e-02  2.55065747e-02\n  -1.92445815e-02 -1.04848109e-03  4.69611622e-02  3.74582224e-03\n   2.17767712e-02 -1.09501882e-02  2.13346910e-02  8.20552837e-03\n   1.53417559e-02  8.27572495e-03  1.02153700e-02 -6.10355437e-02\n  -2.78653782e-02  1.15664331e-02 -1.03326980e-02 -1.80160310e-02\n  -1.45386835e-03  6.07525520e-02 -3.53630702e-03 -1.53880753e-03\n   4.59406190e-02  2.04139631e-02  3.35457288e-02  7.19508296e-03\n   5.07336669e-02 -2.51075961e-02 -7.88376573e-03  8.70501548e-02\n  -1.38269169e-02  1.83919948e-02  1.69022363e-02  1.39089599e-02\n  -5.32541284e-03 -9.55001730e-03  1.91784352e-02 -1.40003888e-02\n  -4.25859205e-02  4.65078391e-02 -1.00065796e-02 -2.39879303e-02\n  -5.01514859e-02 -1.59259478e-03  8.04426800e-03 -6.54568523e-02\n  -3.70679772e-03  2.03577075e-02  1.29384315e-02 -1.40880020e-02\n   1.57173853e-02 -9.11556482e-02  4.22072522e-02 -1.10125206e-02\n   2.85274815e-02  3.14906016e-02 -3.05609293e-02  1.69927515e-02\n   2.64195837e-02 -2.13358086e-02  1.54819814e-02  1.50634916e-02\n  -5.23733906e-03 -1.14251999e-02 -6.18851918e-04 -7.18819350e-02\n  -3.50364335e-02  2.59870123e-02 -2.76175868e-02  9.06186004e-04\n  -1.86457019e-03  6.16583936e-02  5.90859652e-02 -6.04711333e-03\n  -4.16075252e-02 -3.72403115e-03 -1.55113908e-02 -5.07680187e-03\n  -3.68503891e-02 -6.29858896e-02  1.04778139e-02 -4.79897633e-02\n  -5.10004237e-02 -1.65270537e-03  3.07098180e-02 -8.41614965e-04\n  -5.12744002e-02 -1.20785395e-02 -1.68753304e-02  3.58492248e-02\n  -9.43715312e-03  1.89496391e-02  7.84834754e-03  2.39759926e-02\n  -3.08901723e-02 -1.12829283e-02  8.31078924e-03  2.38283519e-02\n   8.61623604e-03  3.41849439e-02 -3.76217030e-02 -1.45408781e-02]]'

In [64]:
train_df['cos_sim'] = train_df['tag_vec_ft'].apply(lambda _df: cosine_similarity(_df, user_tag_vectors))



KeyboardInterrupt: 

In [32]:
k = list(map(float, train_df['tag_vec_ft'][0]))#.astype(float)

ValueError: could not convert string to float: '['

In [49]:
b = ['가', [[-31232.123]], 312]
c = ['나', [[-2.123]], 44]
a = pd.DataFrame([b, c])#, axis=1)
a

0               1    2
0  가  [[-31232.123]]  312
1  나      [[-2.123]]   44

In [50]:
a.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       2 non-null      object
 1   1       2 non-null      object
 2   2       2 non-null      int64 
dtypes: int64(1), object(2)
memory usage: 176.0+ bytes


In [46]:
a.to_csv('a.csv', encoding='utf-8-sig')

In [47]:
b = pd.read_csv('a.csv')
b

Unnamed: 0  0    1    2
0           0  가  212  312
1           1  나   23   44

In [48]:
b.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  2 non-null      int64 
 1   0           2 non-null      object
 2   1           2 non-null      int64 
 3   2           2 non-null      int64 
dtypes: int64(3), object(1)
memory usage: 192.0+ bytes
